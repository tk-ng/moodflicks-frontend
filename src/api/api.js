import axios from "axios";

const baseURL = import.meta.env.VITE_BASE_URL || "http://localhost:3001/";

/** API Class.
 *
 * Static class tying together methods used to get/send to to the API.
 * There shouldn't be any frontend-specific stuff here, and there shouldn't
 * be any API-aware stuff elsewhere in the frontend.
 *
 */

export class Api {
	// the token for interactive with the API will be stored here.
	static token;

	static async request(url, data = {}, method = "get") {
		// console.debug("API Call:", url, data, method);

		const headers = { Authorization: `Bearer ${Api.token}` };
		const params = method === "get" ? data : {};

		try {
			return (
				await axios({
					url,
					method,
					baseURL,
					data,
					params,
					headers,
					allowAbsoluteUrls: false,
				})
			).data;
		} catch (err) {
			console.error("API Error:", err.response);
			let message = err.response.data.error.message;
			throw Array.isArray(message) ? message : [message];
		}
	}

	// Individual API routes

	static async getUser(username) {
		let res = await this.request(`/users/${username}`);
		return res.user;
	}

	static async saveChanges({ username, data }) {
		let res = await this.request(
			`/users/${username}`,
			{ username, data },
			"patch"
		);
		return res.user;
	}

	static async deleteUser(username) {
		let res = await this.request(`/users/${username}`, {}, "delete");
		return res.user;
	}

	static async login(data) {
		let res = await this.request("auth/token", data, "post");
		return res.token;
	}

	static async signup(data) {
		let res = await this.request("auth/register", data, "post");
		return res.token;
	}

	static async getTrendingMovies() {
		let res = await this.request("/movies");
		return res.movies;
	}

	static async searchMovie(keyword) {
		let res = await this.request(`/movies/search/?keyword=${keyword}`);
		return res.movies;
	}

	static async getMovieDetail(id) {
		let res = await this.request(`/movies/${id}`);
		return res.movie;
	}

	static async getMovieCredits(id) {
		let res = await this.request(`/movies/${id}/credits`);
		return res.credits;
	}

	static async getActivities() {
		let res = await this.request(`/moods/activities`);
		return res.activities;
	}

	static async searchMood(keyword) {
		let res = await this.request(`/moods/search/?keyword=${keyword}`);
		return res.moods;
	}

	static async getMoods() {
		let res = await this.request(`/moods`);
		return res.moods;
	}

	static async getMoodsByUser(username) {
		let res = await this.request(`/users/${username}/moods`);
		return res.moods;
	}

	static async getMood(id) {
		let res = await this.request(`/moods/${id}`);
		return res.mood;
	}

	static async getMoviesByMood(id) {
		let res = await this.request(`/moods/${id}/movies`);
		return res.movies;
	}

	static async getMoodsByMovie(id) {
		let res = await this.request(`/movies/${id}/moods`);
		return res.moods;
	}

	static async getSimilarMoods(id) {
		let res = await this.request(`/moods/${id}/similar`);
		return res.moods;
	}

	static async getRecommendedMoviesByMood(id) {
		let res = await this.request(`/moods/${id}/recommended`);
		return res.movies;
	}

	static async createMood({ mood, username }) {
		let res = await this.request(`/moods`, { mood, username }, "post");
		return res.mood;
	}

	// needs username to verify in backend
	static async saveMoodChanges({ id, name, username }) {
		let res = await this.request(`/moods/${id}`, { name, username }, "patch");
		return res.message;
	}

	// needs username to verify in backend
	static async addMovieToMood({ moodId, movieId, username }) {
		const movies = await this.getMoviesByMood(moodId);

		if (movies.length) {
			const movieIds = movies.map((m) => m.id);
			if (movieIds.includes(movieId)) return null;
		}

		let res = await this.request(
			`/moods/${moodId}/movies/${movieId}`,
			{ username },
			"post"
		);
		return res.movieMood;
	}

	// needs username to verify in backend
	static async removeMovieFromMood({ moodId, movieId, username }) {
		let res = await this.request(
			`/moods/${moodId}/movies/${movieId}`,
			{ username },
			"delete"
		);
		return res.message;
	}

	// needs username to verify in backend
	static async deleteMood({ id, username }) {
		let res = await this.request(`/moods/${id}`, { username }, "delete");
		return res.message;
	}

	static async voteMood({ moodId, voteType, username }) {
		let res = await this.request(
			`/moods/${moodId}/vote/${voteType}`,
			{ username },
			"post"
		);
		return res.message;
	}
}
