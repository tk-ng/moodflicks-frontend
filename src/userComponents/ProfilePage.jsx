import { useContext, useEffect, useState } from "react";
import { UserContext } from "./UserContext";
import { Navigate, useParams } from "react-router";
import { Api } from "../api/api";
import { Button } from "reactstrap";
import { useNavigate } from "react-router";
import { MoodList } from "../moodComponents/MoodList";
import { toast } from "react-toastify";

export const ProfilePage = () => {
	const params = useParams();
	const navigate = useNavigate();
	const [user, setUser] = useState();
	const [moods, setMoods] = useState([]);
	const { currUser } = useContext(UserContext);

	const [isLoading, setIsLoading] = useState(true);
	const [notFound, setNotFound] = useState(false);

	useEffect(() => {
		async function getUserInfo(username) {
			try {
				setIsLoading(true);
				setNotFound(false);
				const [userResult, moodsResult] = await Promise.all([
					Api.getUser(username),
					Api.getMoodsByUser(username),
				]);
				setUser(userResult);
				setMoods(moodsResult);
			} catch (err) {
				console.error("User not found:", err);
				setNotFound(true);
			} finally {
				setIsLoading(false);
			}
		}

		getUserInfo(params.username);
	}, [params.username]);

	if (isLoading) return <p>Loading...</p>;
	if (notFound) {
		toast.warning(`username: ${params.username} could not be found.`);
		return <Navigate to="/" />;
	}

	return (
		<div className="text-light">
			<span className="display-6">{user.username}</span>
			{currUser && user.username === currUser.username && (
				<Button
					onClick={() => navigate(`/users/${user.username}/edit`)}
					className="m-3 "
				>
					Edit Profile
				</Button>
			)}
			<div>
				{moods.length ? (
					<>
						<h3> Mood-based movie lists, curated by {user.username}</h3>
						<MoodList moods={moods} hideCreatedBy={true} />
					</>
				) : (
					<span>{user.username} has not created any mood lists yet.</span>
				)}
			</div>
		</div>
	);
};
