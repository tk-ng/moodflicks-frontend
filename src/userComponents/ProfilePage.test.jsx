import { render, screen, waitFor } from "@testing-library/react";
import { Api } from "../api/api";
import { ProfilePage } from "./ProfilePage";
import { useParams, MemoryRouter } from "react-router";
import { UserContext } from "./UserContext";

vi.mock("../api/api", () => ({
	Api: {
		getUser: vi.fn(),
		getMoodsByUser: vi.fn(),
	},
}));

vi.mock("react-router", async function () {
	const mod = await vi.importActual("react-router");
	return {
		...mod,
		useParams: vi.fn(),
	};
});

describe("Renders the ProfilePage component", function () {
	useParams.mockReturnValue({ username: "testuser" });
	it("mounts without crashing", async function () {
		const { container, asFragment } = render(
			<MemoryRouter>
				<UserContext.Provider value={{ currUser: null }}>
					<ProfilePage />
				</UserContext.Provider>
			</MemoryRouter>
		);
		expect(container).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});
});

describe("Display/Hide Edit Profile button based on logged in status", function () {
	Api.getUser.mockResolvedValue({
		username: "testuser",
		firstName: "U1F",
		lastName: "U1L",
		email: "testuser@test.com",
		isAdmin: false,
	});
	Api.getMoodsByUser.mockResolvedValue([
		{
			movies: ["testid1", "testid2", "testid3", "testid4"],
			count: 4,
			mood: "testmood-1",
			id: "8",
			createdBy: "testuser2",
			totalVotes: "2",
		},
		{
			movies: ["testid5", "testid6", "testid7", "testid8", "testid9"],
			count: 5,
			mood: "testmood-2",
			id: "11",
			createdBy: "testuser1",
			totalVotes: "0",
		},
	]);

	it("renders the ProfilePage properly for profile user", async function () {
		render(
			<MemoryRouter>
				<UserContext.Provider value={{ currUser: { username: "testuser" } }}>
					<ProfilePage />
				</UserContext.Provider>
			</MemoryRouter>
		);
		await waitFor(() =>
			expect(
				screen.getByText("Edit Profile", { selector: "button" })
			).toBeInTheDocument()
		);
		expect(screen.getByText("testmood-1")).toBeInTheDocument();
		expect(screen.getByText("testmood-2")).toBeInTheDocument();

		// moodlists on a user's profile page should not include the "created by: " text
		expect(
			screen.queryByText("created by: testuser1", { exact: false })
		).not.toBeInTheDocument();
		expect(
			screen.queryByText("created by: testuser2", { exact: false })
		).not.toBeInTheDocument();
	});

	it("renders the ProfilePage properly for non-profile user", async function () {
		render(
			<MemoryRouter>
				<UserContext.Provider
					value={{ currUser: { username: "noProfileUser" } }}
				>
					<ProfilePage />
				</UserContext.Provider>
			</MemoryRouter>
		);
		await waitFor(() =>
			expect(screen.getByText("testmood-1")).toBeInTheDocument()
		);

		expect(
			screen.queryByText("Edit Profile", { selector: "button" })
		).not.toBeInTheDocument();
	});

	it("renders the ProfilePage properly for anon user", async function () {
		render(
			<MemoryRouter>
				<UserContext.Provider value={{ currUser: null }}>
					<ProfilePage />
				</UserContext.Provider>
			</MemoryRouter>
		);
		await waitFor(() =>
			expect(screen.getByText("testmood-1")).toBeInTheDocument()
		);

		expect(
			screen.queryByText("Edit Profile", { selector: "button" })
		).not.toBeInTheDocument();
	});
});
