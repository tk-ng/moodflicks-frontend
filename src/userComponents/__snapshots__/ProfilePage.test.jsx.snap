// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders the ProfilePage component > mounts without crashing 1`] = `
<DocumentFragment>
  <p>
    Loading...
  </p>
</DocumentFragment>
`;
